### ISSUE TITLE
Merge the Insight patches for zcashd into the main repo

### ISSUE THREAD

**Issue by daira** [zcash/zcash#3708]
The fact that the [patched-for-Insight version of zcashd](https://github.com/zcash-hackworks/zcash-patched-for-explorer) is maintained separately results in it being often out-of-date. After both the Overwinter and Sapling upgrades, this caused Insight-based block explorers to go down until they could be updated. This ticket would merge these patches into the main repo.

I've looked at the diff between the patched-for-Insight version and current master. The differences are:

* `CDBWrapper` has support for enabling leveldb compression and raising the open files limit. These are controlled by two new options (not documented in the help):
  * `-dbmaxopenfiles` (default 1000; current zcashd uses 64)
  * `-dbcompression` (default enabled; current zcashd has it disabled)
* there are three additional indexing options:
  * `-addressindex`: Maintain a full address index, used to query for the balance, txids and unspent outputs for addresses (default: 0)
  * `-timestampindex`: Maintain a timestamp index for block hashes, used to query blocks hashes by a range of timestamps (default: 0)
  * `-spentindex`: Maintain a full spent index, used to query the spending txid and input index for an outpoint (default: 0);
* the computation of the `nBlockTreeDBCache` size is different when either `-addressindex` or `-spentindex` are enabled;
* trivial: there are additional logging lines for the block index database configuration;
* there are several new RPC calls:
  * `getaddressbalance`: Returns the balance for an address(es) (requires addressindex to be enabled).
  * `getaddressdeltas`: Returns all changes for an address (requires addressindex to be enabled).
  * `getaddressmempool`: Returns all mempool deltas for an address (requires addressindex to be enabled).
  * `getaddresstxids`: Returns the txids for an address(es) (requires addressindex to be enabled).
  * `getaddressutxos`: Returns all unspent outputs for an address (requires addressindex to be enabled).
  * `getblockhashes`: Returns array of hashes of blocks within the timestamp range provided. (undocumented dependency on `-timestampindex`)
  * `getspentinfo`: Returns the txid and index where an output is spent. (undocumented dependency on `-spentindex`)
* `getrawtransaction` is changed; I don't quite see what the change is doing from the diff;
* `CScript` is extended with a method to check for a P2PKH script;
* there is a new executable called `wallet-utility` (conditional on `ENABLE_WALLET`). The help for this utility is:
  ```
  This program outputs Bitcoin addresses and private keys from a wallet.dat file
  
  Usage and options:
   -datadir=<directory> to tell the program where your wallet is
   -wallet=<name> (Optional) if your wallet is not named wallet.dat
   -regtest or -testnet (Optional) dumps addresses from regtest/testnet
   -dumppass (Optional)if you want to extract private keys associated with addresses
      -pass=<walletpassphrase> if you have encrypted private keys stored in your wallet
  ```
* there are additional tests for the new indices and RPC calls, and for the `wallet-utility` executable;
* in the RPC test framework, some IPv6-related code is made conditional on a check for local IPv6 support (i.e. whether there is a route to IPv6 localhost).

`-addressindex` appears to only support t-addresses. The additional code to support the indices, their serialization in leveldb, and the new RPC calls is verbose but not too excessive, although I would suggest some of it should be moved out of `main.cpp`, `main.h`, and `rpc/misc.cpp`. I would also suggest some minor refactoring to slightly reduce overhead when the indices are not enabled (although maybe this is microoptimization).

Obviously the `-dbmaxopenfiles` and `-dbcompression` options would have to be documented, as would the dependencies of the new `getblockhashes` and `getspentinfo` RPC calls on the indexing options. `-dbcompression` would have to be default-disabled to maintain block database compatibility. Other than that, there is nothing I can see that would be an obstacle to merging.

**Comment by daira** [comment]
The total amount of work needed for refactoring the patches into clean PRs and reviewing them is fairly substantial (although we've integrated much larger patches before); probably a week's work for someone. I guess it depends how highly we prioritize having Insight explorers working continuously. It *would* make it easier to fix any bugs in the explorers.

**Comment by radix42** [comment]
The botcore patched version of zcashd that I’m using on my block explorers,
and other Insight instances are too is at:

https://github.com/wo01/zcash/tree/bitcore

It’s at 2.0.0 and works perfectly with the Insight code in the
zcash-hackworks repo, while the patched zcashd 2.0.1 there does not (iirc
it’s the one running on the testnet explorer)

I haven’t run a diff between the two branches to see what might be breaking
some features

**Comment by LarryRuane** [comment]
That repo (https://github.com/wo01/zcash/tree/bitcore) now seems to be updated to 2.0.2! I'm testing it now to see if the blocks api is fixed. I would be glad to take a try at doing this work (as outlined by Daira in the first comment), if there are no objections. (I'm currently assigned related #3701.)

**Comment by renuzit** [comment]
Let me know if I can help out with this issue! I'm only a novice c++ dev, but I have tons of experience with insight explorer and JavaScript.

**Comment by LarryRuane** [comment]
Thank you, @renuzit, I very much would appreciate your help! Pretty soon, all the changes to support block explorer will be in zcash master, and then, for one thing, we need to test the insight explorer (which is still at https://github.com/zcash-hackworks), with the updated zcashd master. If you could help with that, it would be a big help (I don't know much about the explorer, just the opposite of you). I think we'd like to move that code into the zcash/zcash repo so that it's officially supported (or at least much more so than today).

**Comment by renuzit** [comment]
@LarryRuane Sure thing. I'll look at it tonight. Do you by chance have a branch with all the zcash insight code merged into it? Or do I need to merge all of the patches one by one?

**Comment by garethtdavies** [comment]
@renuzit see https://github.com/zcash-hackworks/zcash-patched-for-explorer/tree/v2.0.3-insight-explorer. Latest branch is 2.0.3. 

If you want to run in a Docker container then see https://github.com/zcash-hackworks/zcash-insight-explorer.

@LarryRuane 2.0.3 will be deprecated really soon (545135). Any plans to update this before it is merged into master? I rebased 2.0.4 here: https://github.com/garethtdavies/zcash-patched-for-explorer/tree/v2.0.4-insight-explorer and replaced my zcashd with it and it seems to work.

```
/home/zcashd/zcash-patched-for-explorer/src/zcash-cli getdeprecationinfo
{
  "version": 2000350,
  "subversion": "/MagicBean:2.0.3/",
  "deprecationheight": 545135
}
```

**Comment by renuzit** [comment]
@garethtdavies It appears that @LarryRuane is working on a rewrite based on the bitcore patches, I'm really asking if there is a branch with all of the new patches merged in for 2.0.5+. Or if I need to go through each PR and rebase manually 🤮

Side note, I've already tried rebasing 2.0.5 with the zcash-patched-for-explorer project, without much luck. There looks to be evidence that there is an issue with the getaddressbalance PRC call, but I'm still working on figuring out what's wrong there. More than likely it works using @LarryRuane 's patches.

**Comment by garethtdavies** [comment]
> @garethtdavies It appears that @LarryRuane is working on a rewrite based on the bitcore patches, I'm really asking if there is a branch with all of the new patches merged in for 2.0.5+. Or if I need to go through each PR and rebase manually 🤮

Right, I assumed (probably incorrectly) that the result of these four outstanding PRs is what was already running in that patched 2.0.3 branch but packaged in a way for easy review.

My other point concerned the immediacy of this in that 2.0.3 will reach EOL before these PRs are merged into master.

**Comment by LarryRuane** [comment]
@renuzit, @garethtdavies, I just rebased this branch: https://github.com/zcash/zcash/pull/3972, to latest zcash master, and that branch has all the Insight block explorer changes. Thank you both for  your help!

But that's a good point about 2.0.3 EOL; maybe I need to rebase https://github.com/zcash-hackworks/zcash-patched-for-explorer to latest master right now. Last time I rebased it, it wasn't difficult (but maybe this time won't be so easy).

**Comment by LarryRuane** [comment]
Update on my previous comment: I should have realized, rebasing zcash-patched-for-explorer to v2.0.5 is extremely difficult to do now, because a large number of block explorer changes have already been merged into zcash v2.0.5. I started the rebase, and there are tons of merge conflicts. Tried something else: I rebased zcash-patched-for-explorer on v2.0.4 instead, and that went smoothly (a few merge conflicts, but easy to resolve). I just pushed that branch to my fork of hackworks: https://github.com/LarryRuane/zcash-patched-for-explorer/tree/v2.0.4-insight-explorer in case you'd like to take a look or test it (I haven't tested it yet).

**Comment by LarryRuane** [comment]
@garethtdavies 
> My other point concerned the immediacy of this in that 2.0.3 will reach EOL before these PRs are merged into master.

It seems 2.0.3 has already hit EOS (EOL) earlier today but only on mainnet; that height is still 2 months away (if my math is correct). The testnet explorer at https://explorer.testnet.z.cash/ is still humming along.

**Comment by garethtdavies** [comment]
> It seems 2.0.3 has already hit EOS (EOL) earlier today but only on mainnet; that height is still 2 months away (if my math is correct). The testnet explorer at https://explorer.testnet.z.cash/ is still humming along.

Looks like https://explorer.testnet.z.cash/ is running 2.0.2 according to the version on the status page of `2000250` which I think reaches EOL at 502708 according to https://github.com/zcash/zcash/blob/8e2ca4d94e2f1bc0835d3d888a969fe2e83e7ece/src/deprecation.h#L11. Is this actually enforced on testnet or just mainnet?

I tried your 2.0.4 branch (on mainnet) with zero issues so it's probably a good idea to have this 2.0.4 branch on the hackworks repo for those running mainnet explorers.

**Comment by LarryRuane** [comment]
I see now that it's only enforced on mainnet: https://github.com/zcash/zcash/blob/8e2ca4d94e2f1bc0835d3d888a969fe2e83e7ece/src/deprecation.cpp#L20

Still, it's good that a 2.0.4 branch is available on hackworks now, and we should upgrade our testnet explorer at https://explorer.testnet.z.cash/ to v2.0.4. I'll try to figure out how to do that. Thanks for your great help, Gareth!

**Comment by LarryRuane** [comment]
@renuzit, (also ping @garethtdavies), please take a look at this large block: 
```
curl https://explorer.testnet.z.cash/api/block/0002f263a6cc41ebb9b07865e457c1b1375f35012b18bfb6cc41a58ed91cf750 | jq .
```
It has 1635 transactions. But https://explorer.testnet.z.cash/block/0002f263a6cc41ebb9b07865e457c1b1375f35012b18bfb6cc41a58ed91cf750 displays only the first 10. Is that by design? The other block explorers either show all of the transactions, for example https://chain.so/block/ZECTEST/0002f263a6cc41ebb9b07865e457c1b1375f35012b18bfb6cc41a58ed91cf750 or include a `Load More` button like https://testnet.zcha.in/blocks/0002f263a6cc41ebb9b07865e457c1b1375f35012b18bfb6cc41a58ed91cf750. 

There should be a way to ask Insight to show all the transactions in a block.

Is this where the 10 limit comes from: https://github.com/zcash-hackworks/insight-api-zcash/blob/9c0ce97ece8a627bfccf3f8f48649669d0721a2e/lib/transactions.js#L254

I could write a ticket for this, but I first wanted to get your take first, since you know Insight. Thanks.

**Comment by garethtdavies** [comment]
@LarryRuane Are you blocking some javascript or any errors there as it seems to load via infinite scroll see this:

![CleanShot 2019-06-11 at 11 44 52](https://user-images.githubusercontent.com/135382/59298197-d8858200-8c3e-11e9-965f-372997f60ba6.gif)

**Comment by LarryRuane** [comment]
I don't know, it must be something like that. I just tried it on Firefox (running in Ubuntu), and it does the demand scrolling. It doesn't scroll on Chrome on both my Ubuntu VM and my windows laptop. I do have JavaScript set to Allowed in my Chrome settings. Thanks for trying that, @garethtdavies, I'm not going to worry about it at least for now.

**Comment by renuzit** [comment]
@LarryRuane that's by design. It's broken up into 10 transactions a page. I'm unsure why your Chrome doesn't scroll, but it would be interesting to see what your JavaScript console says.

**Comment by alexeirbv** [comment]
Hi! What about patched zcashd v2.0.5? It seems that v2.0.4 hit EOL :(

**Comment by Eirik0** [comment]
Closing this ticket as all of the insight related PRs have been merged.
